src/accelerate/utils/fsdp_utils.py

```diff
@@ -32,7 +33,31 @@
 logger = get_logger(__name__)
 
 
-def save_fsdp_model(fsdp_plugin, accelerator, model, output_dir, model_index=0):
+def _is_peft_model(model):
+    if is_peft_available():
+        from peft import PeftModel
+    return is_peft_available() and isinstance(extract_model_from_parallel(model), PeftModel)
+
+
+def _get_model_state_dict(model, adapter_only=False):
+    if adapter_only and _is_peft_model(model):
+        from peft import get_peft_model_state_dict
+
+        return get_peft_model_state_dict(model, adapter_name=model.active_adapter)
+    else:
+        return model.state_dict()
+
+
+def _set_model_state_dict(model, state_dict, adapter_only=False):
+    if adapter_only and _is_peft_model(model):
+        from peft import set_peft_model_state_dict
+
+        return set_peft_model_state_dict(model, state_dict, adapter_name=model.active_adapter)
+    else:
+        return model.load_state_dict(state_dict)
```

Comment:
I'm not the biggest fan of this pattern we're working with here. 

These should live inside of one function, and currently from what I can tell we have 2 different ways of getting the state_dict inside accelerate now:
1. Here
2. `Accelerator.get_state_dict`

This hints at me that perhaps we need to chunk up `Accelerator`'s `get_state_dict` into something that can be called elsewhere (and stored in `utils.modeling` probably). 

Same with `_set_state_dict` as well. While yes it's just FSDP, then we can write a check for that if necessary. 

Secondly: we generally don't follow the practice of hidden function names in Accelerate, and everything should be made public unless for *extreme* circumstances, and I'm not convinced this is one of those. 

Can we rewrite this a bit to make it more extensible? 

Perhaps just do an `if/else` for calling `set_peft_model_state_dict` or not, and just importing them at the top of the file. This way we can avoid this entirely.

Reply:
@muellerzr See the newest refactor, but I believe this is not possible.

`Accelerator.get_state_dict` already has logic for FSDP, but it handles a different use case for getting the FSDP state dict. `Accelerator.get_state_dict` always returns the `FULL_STATE_DICT` from FSDP. Meanwhile, this file, gets the state dict according to how the user configured they want it to be *saved* (could be a `SHARDED_STATE_DICT`). I believe that is why this file originally never called `Accelerator.get_state_dict`.

As for your suggestion of importing `get_peft_model_state_dict` at the top v.s. encapsulated in the hidden function, we can't import at the top because `accelerate` does not have a dependency on `peft` and importing at the top would throw an error if `peft` is not available on a users computer. Also, the hidden function `_get_model_state_dict` is used 3 times in this file. I could get rid of the hidden function and replace with an inline if-statement in those 3 places, but it would introduce 3 places of repeated code that would need to be updated together. Between the repeated code and the need for keeping the `peft` imports local v.s. top-of-file imports, I think it may make sense to keep the two hidden functions here.

Reply:
@muellerzr Have you had a chance to look this over? Waiting on this PR to get https://github.com/huggingface/transformers/pull/28297 merged.